packages/payment-detection/src/erc777/superfluid-retriever.ts

```diff
@@ -92,41 +89,79 @@ export class SuperFluidInfoRetriever {
+          continue;
+        }
+        const amount =
+          diffFlowRate * (streamEvents[index].timestamp - streamEvents[index - 1].timestamp);
+        paymentEvents.push({
+          amount: amount.toString(),
+          name: this.eventName,
+          parameters: {
+            to: this.toAddress,
+            block: parseInt(streamEvents[index].blockNumber.toString()),
+            txHash: streamEvents[index].transactionHash,
+            streamEventName: getEventName(streamEvents[index]),
+          },
+          timestamp: streamEvents[index].timestamp,
+        });
       }
-      const amount =
-        diffFlowRate * (streamEvents[index].timestamp - streamEvents[index - 1].timestamp);
-      paymentEvents.push({
-        amount: amount.toString(),
+      if (paymentEvents.length > 0) {
+        const newLastParameters = paymentEvents[paymentEvents.length - 1].parameters;
+        if (lastEventOngoing && newLastParameters) {
+          newLastParameters.streamEventName = PaymentTypes.STREAM_EVENT_NAMES.START_STREAM;
+          paymentEvents[paymentEvents.length - 1].parameters = newLastParameters;
+        }
+      }
+    }
+
+    // Transform the one off payment events into ERC777PaymentNetworkEvent
+    const oneOffPaymentEventsUpdated = oneOffPaymentEvents.map((event) => {
+      return {
+        amount: event.amount.toString(),
         name: this.eventName,
         parameters: {
           to: this.toAddress,
-          block: parseInt(streamEvents[index].blockNumber.toString()),
-          txHash: streamEvents[index].transactionHash,
-          streamEventName: getEventName(streamEvents[index]),
+          block: parseInt(event.blockNumber.toString()),
+          txHash: event.transactionHash,
+          // This is always the case:
+          //  - If no previous payment event we can consider it an an END_STREAM event
+          //  - If last payment event is END_STREAM this one is as well
+          //  - If last payment event is START_STREAM, the last payment event after sorting will be of type START_STREAM,
+          //    so we can consider this one as END_STREAM
+          streamEventName: PaymentTypes.STREAM_EVENT_NAMES.END_STREAM,
         },
-        timestamp: streamEvents[index].timestamp,
-      });
-    }
-    if (paymentEvents.length > 0) {
-      const newLastParameters = paymentEvents[paymentEvents.length - 1].parameters;
-      if (lastEventOngoing && newLastParameters) {
-        newLastParameters.streamEventName = PaymentTypes.STREAM_EVENT_NAMES.START_STREAM;
-        paymentEvents[paymentEvents.length - 1].parameters = newLastParameters;
-      }
-    }
-    return paymentEvents;
+        timestamp: event.timestamp,
+      } as PaymentTypes.ERC777PaymentNetworkEvent;
+    });
+
+    // Insert the transformed one off payment events into the payments events, sort them and update their streamEventName parameter
+    return paymentEvents.concat(oneOffPaymentEventsUpdated).sort((a, b) => {
```

Comment:
This way of adding a custom sort directly on the return is a bit hard to read, I think. Maybe for another PR but the 2 different logics to handle 2 different kinds of events could be extracted to sub-functions to make unit tests easier.

Reply:
Having two sub-functions is a good idea for readability purposes. But for the unit tests we should test the global function as we have to check that the all events are ordered correctly,